nutils/evaluable.py

```diff
@@ -3715,11 +3765,14 @@ def _simplified(self):
       return zeros_like(self)
     elif self.index not in self.func.arguments:
       return Ravel(Transpose.from_end(InsertAxis(self.func, self.index.length), -2))
-    for iaxis, axis in enumerate(self.func._axes[:-1]):
-      if isinstance(axis, Inserted) and self.index not in axis.length.arguments:
-        return insertaxis(loop_concatenate(self.func._uninsert(iaxis), self.index), iaxis, axis.length)
-    if isinstance(self.func._axes[-1], Inserted) and self.index not in self.func.shape[-1].arguments and not equalindex(self.func.shape[-1], 1):
-      return Ravel(InsertAxis(loop_concatenate(InsertAxis(self.func._uninsert(self.func.ndim-1), 1), self.index), self.func.shape[-1]))
+    unaligned, where = unalign(self.func)
+    for iaxis, length in enumerate(self.func.shape):
+      if iaxis not in where and self.index not in length.arguments:
+        uninserted = align(unaligned, [i-(i>iaxis) for i in where], self.func.shape[:iaxis]+self.func.shape[iaxis+1:])
+        if iaxis != self.ndim-1:
+          return insertaxis(loop_concatenate(uninserted, self.index), iaxis, length)
+        elif not equalindex(length, 1):
+          return Ravel(InsertAxis(loop_concatenate(InsertAxis(uninserted, 1), self.index), length))
```

Comment:
It was not entirely trivial but I came up with the following:
```python
unaligned, where = unalign(self.func)
if self.ndim-1 not in where:
  # reinsert concatenation axis, at unit length if possible so we can
  # insert the remainder outside of the loop
  unaligned = InsertAxis(unaligned, self.func.shape[-1] if self.index in self.func.shape[-1].arguments else 1)
  where += self.ndim-1,
elif where[-1] != self.ndim-1:
  # bring concatenation axis to the end
  unaligned = Transpose(unaligned, numpy.argsort(where))
  where = tuple(sorted(where))
f = loop_concatenate(unaligned, self.index)
if not equalindex(self.shape[-1], f.shape[-1]):
  # last axis was reinserted at unit length AND it was not unit length
  # originally - if it was unit length originally then we proceed only if
  # there are other insertions to promote, otherwise we'd get a recursion.
  f = Ravel(InsertAxis(f, self.func.shape[-1]))
elif len(where) == self.ndim:
  return
return align(f, where, self.shape)
```